### PR TITLE
7/add thermometer

### PR DESCRIPTION
**Pull request by hagoptaminian**

### CODE REVIEW

File: src/utils/canvas.js [hunk 1]
```diff
@@ -155,24 +155,164 @@ export const generateHalfMountainPoints = (
...
+        (2 * index + 1) * centralAtomRadius +
+        index * spaceBetweenMolecules,
+    );
+  return centerXs;
+};
+
+export const distributeMoleculesRandomly = (moleculeDistribution) => {
+  const numberOfMolecules = determineNumberOfMolecules(moleculeDistribution);
+  const randomDistribution = new Array(numberOfMolecules);
+  const moleculeDistributionArray = Object.entries(moleculeDistribution);
```
Comment (pyphilia) [inline src/utils/canvas.js:262]: `_.shuffle()` might be the solution you were looking for? This is interesting too https://flaviocopes.com/how-to-shuffle-array-javascript/

**Review by pyphilia — CHANGES_REQUESTED** [review]
Nice PR! Here some comments.

File: src/components/lab/canvas/sky/greenhouse-gases/MoleculesRow.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  MOLECULE_DISTRIBUTION,
+  X_DISTANCE_BETWEEN_MOLECULES_IN_ROW,
+} from '../../../../../config/constants';
+import {
+  determineMoleculeCenterXs,
+  distributeMoleculesRandomly,
+} from '../../../../../utils/canvas';
+import Molecule from './Molecule';
+
+const MoleculesRow = ({ skyHeight, skyWidth, skyBeginsX, skyBeginsY }) => {
```
[the remark this answers is not in the document]
Comment (hagoptaminian) [reply src/components/lab/canvas/sky/greenhouse-gases/MoleculesRow.js:16]: Interesting suggestion! My thinking is:

- It is helpful to be explicit which height/width we are talking about. A generic name like 'height' or 'width' would be confusing. For example, is it the canvas height or the sky height?
- There are so many x's and y's throughout the application that, again, it is helpful to spell out explicitly what the x's and y's are.
- It is very unlikely that the MoleculeRow will be moved from the sky - and even if it is, adapting it to the pattern of the other canvas components (Sea, Ground, etc) is quite straightforward.

So I will keep as is...

File: src/components/lab/canvas/sky/greenhouse-gases/MoleculesRow.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  const moleculeRowBeginsX = skyBeginsX + MOLECULE_ROW_BEGINS_X * skyWidth;
+  const moleculeRowBeginsY = skyBeginsY + MOLECULE_ROW_BEGINS_Y * skyHeight;
+  const centralAtomRadius = ATOM_DIMENSIONS.medium * skyWidth;
+  const spaceBetweenMolecules = X_DISTANCE_BETWEEN_MOLECULES_IN_ROW * skyWidth;
+
+  // determine a random molecule distribution so that the molecules are not all shown grouped together by type
+  // store molecule distribution in a ref so that the distribution isn't re-computed when screen size changes
+  // TBD: If/how to modify this when we allow for MOLECULE_DISTRIBUTION to change via redux (e.g. 'add more methane')
+  const randomMoleculeDistribution = useRef(
```
[the remark this answers is not in the document]
Comment (hagoptaminian) [reply src/components/lab/canvas/sky/greenhouse-gases/MoleculesRow.js:25]: So basically `useRef` creates an immutable object that keeps the same reference between renders. If we just used `const randomMoleculeDistribution = distributeMoleculesRandomly(MOLECULE_DISTRIBUTION)`, then whenever the screen is resized, `distributeMoleculesRandomly` is recalculated, and the molecules jump around, which we don't want. With `useRef`, these value changes don't cause a re-render.

But I think combining `useState` and `useEffect` can achieve the same result, and is the more common approach, so I will revert to that.

File: src/components/lab/canvas/sky/greenhouse-gases/Molecule.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      );
+      break;
+    case WATER:
+      moleculeToDisplay = (
+        <Water moleculeCenter={moleculeCenter} skyWidth={skyWidth} />
+      );
+      break;
+    case METHANE:
+      moleculeToDisplay = (
+        <Methane moleculeCenter={moleculeCenter} skyWidth={skyWidth} />
```
[the remark this answers is not in the document]
Comment (hagoptaminian) [reply src/components/lab/canvas/sky/greenhouse-gases/Molecule.js:27]: Very cool. Did not know one could do this.

**Pull request merged**